### ISSUE TITLE
/data/files_external/rootcerts.crt empty after update

### ISSUE THREAD

**Issue by LukasReschke** [nextcloud/server#4978]
https://github.com/nextcloud/server/issues/3470 is back but this time it happens every time after an update via the web updater on c.nc.c.

**Comment by LukasReschke** [comment]
Looks really odd, it claims the file hasn't been changed since 1475071065 which would be Wed, 28 Sep 2016 13:57:45 GMT 😱 
```
mysql> select * from oc_filecache where name like "%rootcerts.crt%";
+--------+---------+------------------------------+----------------------------------+--------+---------------+----------+----------+--------+------------+---------------+-----------+------------------+----------------------------------+-------------+----------+
| fileid | storage | path                         | path_hash                        | parent | name          | mimetype | mimepart | size   | mtime      | storage_mtime | encrypted | unencrypted_size | etag                             | permissions | checksum |
+--------+---------+------------------------------+----------------------------------+--------+---------------+----------+----------+--------+------------+---------------+-----------+------------------+----------------------------------+-------------+----------+
|  25878 |      11 | files_external/rootcerts.crt | 96f0eb1b0b28e846975c98dbeb5e734d |  13459 | rootcerts.crt |        7 |        3 | 757353 | 1475071065 |    1475071065 |         0 |                0 | f91f134cfa8930a8a5755af742d03c64 |          27 |          |
+--------+---------+------------------------------+----------------------------------+--------+---------------+----------+----------+--------+------------+---------------+-----------+------------------+----------------------------------+-------------+----------+
1 row in set (2.95 sec)
```

**Comment by LukasReschke** [comment]
The following error log is also appearing around those times:

> {"reqId":"H0v9Bq3mBEShyjFr1N71","level":3,"time":"2017-05-19T20:12:48+00:00","remoteAddr":"","user":"--","app":"PHP","method":"--","url":"--","message":"filemtime(): stat failed for \/var\/www\/html\/resources\/config\/ca-bundle.crt at \/var\/www\/html\/lib\/private\/Security\/CertificateManager.php#275","userAgent":"--","version":"12.0.0.23"}

**Comment by LukasReschke** [comment]
Looking at the server logs when those cases happened I think I see the following similarities:

1. Cron seems to have run around that time
2. There are federated shares on this instance
3. The following log messages are in the server log and seem to be from cron:

```
{"reqId":"R0CpsfO22v1iKNCZWWDd","level":3,"time":"2017-05-21T21:00:43+00:00","remoteAddr":"","user":"--","app":"PHP","method":"--","url":"--","message":"include(\/var\/www\/html\/3rdparty\/composer\/..\/icewind\/streams\/src\/CallbackWrapper.php): failed to open stream: No such file or directory at \/var\/www\/html\/lib\/composer\/composer\/ClassLoader.php#444","userAgent":"--","version":"12.0.0.27"}
{"reqId":"R0CpsfO22v1iKNCZWWDd","level":3,"time":"2017-05-21T21:00:43+00:00","remoteAddr":"","user":"--","app":"PHP","method":"--","url":"--","message":"include(): Failed opening '\/var\/www\/html\/3rdparty\/composer\/..\/icewind\/streams\/src\/CallbackWrapper.php' for inclusion (include_path='\/var\/www\/html\/apps\/mail\/vendor\/pear-pear.horde.org\/Horde_Util:\/var\/www\/html\/apps\/mail\/vendor\/pear-pear.horde.org\/Horde_Exception:\/var\/www\/html\/apps\/mail\/vendor\/pear-pear.horde.org\/Horde_Stream_Filter:\/var\/www\/html\/apps\/mail\/vendor\/pear-pear.horde.org\/Horde_Stream:\/var\/www\/html\/apps\/mail\/vendor\/pear-pear.horde.org\/Horde_Socket_Client:\/var\/www\/html\/apps\/mail\/vendor\/pear-pear.horde.org\/Horde_Stream_Wrapper:\/var\/www\/html\/apps\/mail\/vendor\/pear-pear.horde.org\/Horde_Secret:\/var\/www\/html\/apps\/mail\/vendor\/pear-pear.horde.org\/Horde_Text_Flowed:\/var\/www\/html\/apps\/mail\/vendor\/pear-pear.horde.org\/Horde_Mail:\/var\/www\/html\/apps\/mail\/vendor\/pear-pear.horde.org\/Horde_ListHeaders:\/var\/www\/html\/apps\/mail\/vendor\/pear-pear.horde.org\/Horde_Translation:\/var\/www\/html\/apps\/mail\/vendor\/pear-pear.horde.org\/Horde_Support:\/var\/www\/html\/apps\/mail\/vendor\/pear-pear.horde.org\/Horde_Crypt_Blowfish:\/var\/www\/html\/apps\/mail\/vendor\/pear-pear.horde.org\/Horde_Nls:\/var\/www\/html\/apps\/mail\/vendor\/pear-pear.horde.org\/Horde_Date:\/var\/www\/html\/apps\/mail\/vendor\/pear-pear.horde.org\/Horde_Mime:\/var\/www\/html\/apps\/mail\/vendor\/pear-pear.horde.org\/Horde_Idna:\/var\/www\/html\/apps\/mail\/vendor\/pear-pear.horde.org\/Horde_Imap_Client:\/var\/www\/html\/apps\/mail\/vendor\/pear-pear.horde.org\/Horde_Smtp:\/var\/www\/html\/apps\/mail\/vendor\/pear-pear.horde.org\/Horde_Text_Filter:\/var\/www\/html\/3rdparty\/pear\/console_getopt:\/var\/www\/html\/3rdparty\/pear\/pear_exception:\/var\/www\/html\/3rdparty\/pear\/archive_tar:\/var\/www\/html\/3rdparty\/pear\/pear-core-minimal\/src:\/var\/www\/html\/apps') at \/var\/www\/html\/lib\/composer\/composer\/ClassLoader.php#444","userAgent":"--","version":"12.0.0.27"}
{"reqId":"R0CpsfO22v1iKNCZWWDd","level":4,"time":"2017-05-21T21:00:44+00:00","remoteAddr":"","user":"--","app":"cron","method":"--","url":"--","message":"Class 'Icewind\\Streams\\CallbackWrapper' not found","userAgent":"--","version":"12.0.0.27"}
```

Let's see if that helps to reproduce this situation somehow.

**Comment by blizzz** [comment]
cron is not supposed to be running in maintenance mode. Or does it? 🙈

**Comment by ChristophWurst** [comment]
> cron is not supposed to be running in maintenance mode. Or does it? 🙈

Also what if we check for maintenance mode, see it's disabled and run cron? Then an update is performed while cron is still running (because it does time consuming tasks). Maybe we should check whether it's running, if that's possible at all.

**Comment by blizzz** [comment]
IIRC it creates a lock file in the data dir.  It already detects the state itself to prevent running in parallel if the last run did not finish.

**Comment by MorrisJobke** [comment]
It creates a dB entry - parallel runs are supported and will double the speed to process a tasks ;)

**Comment by blizzz** [comment]
doh … 👴 … thanks for dating me up @MorrisJobke

**Comment by MorrisJobke** [comment]
> doh … 👴 … thanks for dating me up @MorrisJobke

Afaik there since 10 or 11. I guess @nickvergessen worked on this 😉

**Comment by LukasReschke** [comment]
Happened again after update:

```
root@cloud:/var/www/data/files_external# ls -lah
total 8.0K
drwxr-xr-x  2 www-data www-data 4.0K Jul 12 14:06 .
drwxrwx--- 34 www-data www-data 4.0K Jun 28 19:25 ..
-rw-r--r--  1 www-data www-data    0 Jul 18 08:08 rootcerts.crt
```

**Comment by LukasReschke** [comment]
I updated via PHAR this time.

**Comment by nickvergessen** [comment]
Some ideas:
* Create a non-changable cert file that is used as a fallback when the other one is empty
* When writing the cert file always write to a new name and change a config when the file is none empty which keeps the current name

**Comment by MorrisJobke** [comment]
Should be fixed by #9562

**Comment by MorrisJobke** [comment]
Potential fix is in #9900

**Comment by MorrisJobke** [comment]
Let's see if this happens again on an upgrade to 14+. For now I will close it.
